src/whir/utils.rs

```diff
@@ -36,42 +36,123 @@ pub const fn workload_size<T: Sized>() -> usize {
+pub fn get_challenge_stir_queries<Challenger, F, EF>(
     domain_size: usize,
     folding_factor: usize,
     num_queries: usize,
-    prover_state: &mut Chal,
+    prover_state: &mut Challenger,
 ) -> Result<Vec<usize>, FiatShamirError>
 where
+    Challenger: ChallengeSampler<EF>,
     F: Field,
     EF: ExtensionField<F>,
 {
-    // Folded domain size = domain_size / 2^folding_factor.
+    // PHASE 1: Compute folded domain parameters
+
+    // Apply WHIR folding: domain_size / 2^folding_factor
+    //
+    // Example: 1M domain → 3 folds → 1M >> 3 = 125K domain
     let folded_domain_size = domain_size >> folding_factor;
 
-    // Number of bits needed to represent an index in the folded domain.
+    // Bits required to index into the folded domain
+    //
+    // Example: 125K domain needs ceil(log_2(125K)) ≈ 17 bits per query
     let domain_size_bits = log2_ceil_usize(folded_domain_size);
 
-    // Sample one integer per query, each with domain_size_bits of entropy.
-    let queries = (0..num_queries)
-        .map(|_| prover_state.sample_bits(domain_size_bits) % folded_domain_size)
-        .sorted_unstable()
-        .dedup()
-        .collect();
+    // PHASE 2: Determine batching strategy
+
+    // The maximum number of bits that can be safely sampled in a single call from the challenger.
+    //
+    // To avoid statistical bias when sampling randomness, the number of bits requested should be
+    // less than the bit length of the challenger's underlying field modulus.
+    let max_bits_per_call = F::bits() - 1;
+
+    // Total entropy needed: num_queries × bits_per_query
+    //
+    // Example: 100 queries × 4 bits = 400 bits total
+    let total_bits_needed = num_queries * domain_size_bits;
+
+    // PHASE 3: Execute sampling
+    let queries = if total_bits_needed > 0 && domain_size_bits > 0 {
+        // Pre-allocate result vector
+        let mut all_queries = Vec::with_capacity(num_queries);
+
+        if total_bits_needed <= max_bits_per_call {
+            // SINGLE BATCH PATH
+            //
+            // All entropy fits in one sponge squeeze - maximum efficiency
+            let all_bits = prover_state.sample_bits(total_bits_needed);
+
+            // Bit mask for extracting domain_size_bits chunks
+            //
+            // Example: 4 bits → mask = 0b1111 = 15
+            let mask = (1 << domain_size_bits) - 1;
+
+            // Extract each query index from the packed bit stream
+            for i in 0..num_queries {
+                // Bit position for query i: i × bits_per_query
+                let start_bit = i * domain_size_bits;
+
+                // Extract bits [start_bit, start_bit + domain_size_bits)
+                let query_bits = (all_bits >> start_bit) & mask;
+
+                // Map raw bits to valid domain index via modular reduction
+                let query_index = query_bits % folded_domain_size;
+                all_queries.push(query_index);
```

Comment:
Again, if `folded_domain_size` is a power of `2`, you don't need the quotient.
```suggestion
                let query_bits = all_bits & mask;
                let all_bits >>= domain_size_bits;
                // Map raw bits to valid domain index via modular reduction
                let query_index = query_bits % folded_domain_size;
                all_queries.push(query_index);
```